python/src/numerous/user_session.py

```diff
@@ -0,0 +1,22 @@
+from importlib import import_module
+from typing import Any, Dict, Optional
+from numerous.user import User
+from abc import ABC, abstractmethod
+
+
+class BaseSession(ABC):
+    def __init__(self):
+        self._cookies: Dict[str, Any] = {}
+        self._user: User = User(self._get_cookies)
+
+    @abstractmethod
+    def _get_cookies(self) -> Dict[str, Any]:
+        raise NotImplementedError("Cookies getter is not implemented")
+
+    @property
+    def cookies(self) -> Dict[str, Any]:
+        return self._get_cookies()
+
+    @property
+    def user(self) -> Optional[Dict[str, Any]]:
+        return self._user
```

Comment:
Instead of using `ABC`s I would favour creating class that gets required functionality with dependency injection.

```python
from typing import Protocol

class CookieGetter(Protocol):
    def cookies(self): -> dict[str, Any]: ...

class Session:
    def __init__(self, cg: CookieGetter):
        self._cg = cg
    
    def _user_info(self):
        cookies = self._cg.get_cookies()
        ...

     @property
     def user(self) -> Optional["User"]:
         user_info = self._user_info()

# and then in streamlit.py fx
class StreamlitCookieGetter:  # no inheritance since we use protocols/structural subtyping/static ducktyping
    def cookies(self) -> dict[str, str]:
        ...

session = user_session.Session(cg=StreamlitCookieGetter())

Reply:
This has been implemented with a Protocol as suggested